core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,653 @@
+    this.parent = parent;
+    this.keyValueEntries = keyValueEntries;
+    cascadesCancellation = true;
+  }
+
+  boolean canBeCancelled() {
+    // A context is cancellable if it cascades from its parent and its parent is
+    // cancellable.
+    return (cascadesCancellation && this.parent != null && this.parent.canBeCancelled());
+  }
+
+  /**
+   * Create a new context which is independently cancellable, the new context will cascade
+   * cancellation from its parent.
+   */
+  public CancellableContext withCancellation() {
+    return new CancellableContext(this);
+  }
+
+  /**
+   * Create a new context which will cancel itself after an absolute deadline expressed as
+   * nanoseconds in the {@link System#nanoTime()} clock.
+   *
+   * <p>It is recommended that callers only use this method when propagating a derivative of
+   * a received existing deadline. When establishing a new deadline {@link #withDeadlineAfter}
+   * is the better mechanism.
+   */
+  public CancellableContext withDeadlineNanoTime(long deadlineNanoTime) {
+    return withDeadlineAfter(Math.max(0, deadlineNanoTime - System.nanoTime()),
+        TimeUnit.NANOSECONDS);
+  }
+
+  /**
+   * Create a new context which will cancel itself after the given {@code duration} from now.
+   */
+  public CancellableContext withDeadlineAfter(long duration, TimeUnit unit) {
+    Preconditions.checkArgument(duration >= 0, "duration must be greater than or equal to 0");
+    Preconditions.checkNotNull(unit, "unit");
+    return new CancellableContext(this, unit.toNanos(duration));
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V> Context withValue(Key<V> k1, V v1) {
+    return new Context(this, new Object[][]{{k1, v1}});
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V1, V2> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2) {
+    return new Context(this, new Object[][]{{k1, v1}, {k2, v2}});
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V1, V2, V3> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2, Key<V3> k3, V3 v3) {
+    return new Context(this, new Object[][]{{k1, v1}, {k2, v2}, {k3, v3}});
+  }
+
+  /**
+   * Create a new context which copies the values of this context but does not propagate it's
+   * cancellation and is it's own independent root for cancellation.
+   */
+  public CancellableContext fork() {
+    return new Context(this).withCancellation();
+  }
+
+  /**
+   * Attach this context to the thread and make it {@link #current}, the previously current context
+   * will be restored when detach is called. It is allowed to attach contexts where
+   * {@link #isCancelled()} is {@code true}.
+   */
+  public void attach() {
```

Comment:
We should have a code sample here with how it's expected to be used where the detach is in a finally.